server/src-lib/Hasura/RQL/DDL/Metadata/Types.hs

```diff
@@ -0,0 +1,434 @@
+                    , cfKey
+                    ]
+
+  parseJSON _ =
+    fail "expecting an Object for TableMetadata"
+
+$(deriveToJSON (aesonDrop 3 snakeCase){omitNothingFields=True} ''TableMeta)
+
+data FunctionsMetadata
+  = FMVersion1 ![QualifiedFunction]
+  | FMVersion2 ![Schema.TrackFunctionV2]
+  deriving (Show, Eq, Lift)
+
+instance ToJSON FunctionsMetadata where
+  toJSON (FMVersion1 qualifiedFunctions) = toJSON qualifiedFunctions
+  toJSON (FMVersion2 functionsV2)        = toJSON functionsV2
+
+data ClearMetadata
+  = ClearMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''ClearMetadata)
+
+instance FromJSON ClearMetadata where
+  parseJSON _ = return ClearMetadata
+
+data ReplaceMetadata
+  = ReplaceMetadata
+  { aqVersion          :: !MetadataVersion
+  , aqTables           :: ![TableMeta]
+  , aqFunctions        :: !(Maybe FunctionsMetadata)
+  , aqRemoteSchemas    :: !(Maybe [AddRemoteSchemaQuery])
+  , aqQueryCollections :: !(Maybe [Collection.CreateCollection])
+  , aqAllowlist        :: !(Maybe [Collection.CollectionReq])
+  } deriving (Show, Eq, Lift)
+$(deriveToJSON (aesonDrop 2 snakeCase){omitNothingFields=True} ''ReplaceMetadata)
+
+instance FromJSON ReplaceMetadata where
+  parseJSON = withObject "Object" $ \o -> do
+    version <- o .:? "version" .!= MVVersion1
+    ReplaceMetadata version
+      <$> o .: "tables"
+      <*> (o .:? "functions" >>= mapM (parseFunctions version))
+      <*> o .:? "remote_schemas"
+      <*> o .:? "query_collections"
+      <*> o .:? "allow_list"
+    where
+      parseFunctions = \case
+        MVVersion1 -> fmap FMVersion1 . parseJSON
+        MVVersion2 -> fmap FMVersion2 . parseJSON
+
+data ExportMetadata
+  = ExportMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''ExportMetadata)
+
+instance FromJSON ExportMetadata where
+  parseJSON _ = return ExportMetadata
+
+data ReloadMetadata
+  = ReloadMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''ReloadMetadata)
+
+instance FromJSON ReloadMetadata where
+  parseJSON _ = return ReloadMetadata
+
+data DumpInternalState
+  = DumpInternalState
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''DumpInternalState)
+
+instance FromJSON DumpInternalState where
+  parseJSON _ = return DumpInternalState
+
+data GetInconsistentMetadata
+  = GetInconsistentMetadata
+  deriving (Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''GetInconsistentMetadata)
+
+instance FromJSON GetInconsistentMetadata where
+  parseJSON _ = return GetInconsistentMetadata
+
+data DropInconsistentMetadata
+ = DropInconsistentMetadata
+ deriving(Show, Eq, Lift)
+$(deriveToJSON defaultOptions ''DropInconsistentMetadata)
+
+instance FromJSON DropInconsistentMetadata where
+  parseJSON _ = return DropInconsistentMetadata
+
+-- Functions enable encoding Metadata types and their dependants to ordered JSON
+
+replaceMetadataToOrdJSON :: ReplaceMetadata -> AO.Value
+replaceMetadataToOrdJSON ( ReplaceMetadata
```

Comment:
Could you maybe add a short comment here about what invariants we maintain about order and why? It isn’t immediately clear to me just from reading the code what matters and what doesn’t.

My understanding is that we’re doing this so that metadata export is deterministic, so if you export the same thing twice, you get the same results. That makes complete sense, but it raises questions like: do we guarantee any consistency about ordering across versions of `graphql-engine`? If so, which invariants? And if not, it would be helpful to explicitly say so just so that people reading/modifying the code don’t worry about it.

Reply:
The deterministic export of metadata is already achieved via #3230. This PR intends to remove fields with `[]`/`null`/defaults in exported metadata. And, the order of keys in an object is "hardcoded" using `Data.Aeson.Ordered` module since there's no way to determine the order of keys in `Data.Aeson`.

The CLI based developer experience needs the exported metadata should be precise and clean. The developer should be able to edit the exported YAML file and apply them. In an offline discussion between me, @coco98 and @arvi3411301 we decided the order of the keys in `ReplaceMetadata`, `TableMeta` etc. and the same followed here. 

> do we guarantee any consistency about ordering across versions of graphql-engine?

We do guarantee the order of keys in the exported metadata. I'll add the necessary description as comments.

I think there are 2 kinds of "order" we could consider. 
1. The order of items in an array. (solved in #3230)
2. The order of keys in an object. (In this PR)

Reply:
Okay, thanks, that helps! I’ve tweaked the comment slightly to explicitly mention the fact that the ordering should be maintained across versions; feel free to change it if you think it should be worded differently.

Your note that the encoding should be roundtrippable got me wondering, though: do you think it would make sense to add some quickcheck tests to test that property?

Reply:
Adding a quick check to test the property makes sense. I'll add them in next commit.

Reply:
> To add a quick check, we need to defined generator `Gen a` for all `ReplaceMetadata` dependent types. A lot of code to write. Is there a way to generate `Gen a` with a simple template haskell or type class deriving?. I'm looking at the `Hedgehog` property testing package.

Good question. Neither QuickCheck nor hedgehog provide anything like that out of the box, since they (somewhat reasonably) point out that the choice of generator for each type is important for getting value out of the tests. That said, I think in many cases, you really do just want to use a generic generator for each field of a record type, for example.

The QuickCheck documentation links to two packages that can do that sort of thing: [testing-feat](http://hackage.haskell.org/package/testing-feat) and [generic-random](http://hackage.haskell.org/package/generic-random); I also found the [generic-arbitrary](http://hackage.haskell.org/package/generic-arbitrary) package. In the hedgehog realm, there’s [hedgehog-generic](http://hackage.haskell.org/package/hedgehog-generic), but I’m skeptical of that package’s utility: hedgehog doesn’t have an `Aribitrary` class, so a generic generator is “deeply” generic. I can’t imagine that being what you want very often.

I don’t actually have a ton of experience using property-based testing for anything non-trivial, so I’m not sure what the right thing to do is here. Maybe @paf31 has some thoughts?

Reply:
I don't have a lot of useful input here, but I think it would be nice to avoid adding many `Arbitrary` instances in the `src-lib` area of the code, especially when they'll probably end up being specific to this sort test. Instead, we can use `genericArbitrary` or some other method of getting values of type `Gen a`, entirely within the test suite, and then use the `Testable a => Testable (Gen a)` instance to pass the composed generator itself to `quickCheck`.

I'm not sure to what extent that is common knowledge or not, but it seemed worth mentioning.